### PR TITLE
Add support for error tracking with `TrotterProduct` template 

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
Adding support to compute the error for a given instance of `qml.TrotterProduct`. This is part of the push on error propagation in Pennylane

**Description of the Change:**
Using the paper ["Theory of Trotter Error with Commutator Scaling"](https://arxiv.org/pdf/1912.08854.pdf) to implement upper-bounds on the additive error for the Suzuki Trotter Product formula. 

- Created an error module: `pennylane/resources/error`, moved the base `error.py` file into this repository (similar changes for the tests)
- Created a `trotter_error.py` file in the error module which contains all of the helper functions along with the implementations for the `_one_norm_error` and `_commutator_error` functions. 
- Updated the abstract `ErrorOperation` to change the `error()` from a property to a method. (associated changes in tests)
- Updated `TrotterProduct` to inherit from `ErrorOperation` and implemented the `error()` method. 
- Added integration tests for the new error method in the TrotterProduct class.

Using this error method, we can benchmark the error in approximating time evolution using the Suzuki-Trotter product formulas:

image

### CODE REVIEW

File: tests/templates/test_subroutines/test_trotter.py [hunk 1]
```diff
@@ -464,19 +466,76 @@ def test_private_scalar(self, order, result):
...
+            op.error(method="commutator", fast=False),
+        ):
+            assert isinstance(computed_error, SpectralNormError)
+            assert qnp.isclose(computed_error.error, expected_error)
+
+    @pytest.mark.all_interfaces
+    @pytest.mark.parametrize(
+        "method, expected_error", (("one-norm", 0.009140625), ("commutator", 8e-3))
+    )
+    @pytest.mark.parametrize("interface", ("autograd", "jax", "torch"))
```
Comment (Jaybsoni) [inline tests/templates/test_subroutines/test_trotter.py:521]: @DSGuala @soranjh, 

This test fails for the tensorflow interface. Not sure how we feel about putting in effort to get it passing, not sure if tf support is high priority?

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -117,6 +117,20 @@
...
+  >>> hamiltonian = qml.dot([1.0, 0.5, -0.25], [qml.X(0), qml.Y(0), qml.Z(0)])
+  >>> op = qml.TrotterProduct(hamiltonian, time=0.01, order=2)
+  >>> op.error(method="one-norm")
+  SpectralNormError(8.039062500000003e-06)
+  >>>
+  >>> op.error(method="commutator")
+  SpectralNormError(6.166666666666668e-06)
+  ```
+
+
```
Comment (Jaybsoni) [inline doc/releases/changelog-dev.md:133]:
```suggestion
```

File: pennylane/resource/error/trotter_error.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+def _generate_combinations(num_variables, required_sum):
+    r"""A helper function which generates a sequence of valid combinations which
+    satisfy the required sum constraint.
+
+    **Example:**
+
+        Suppose you have :math:`k` variables :math:`{a_1, a_2, ..., a_k}`. How many unique combinations
+        of positive integers are there that satisfy the constraint :math:`\Sum_{i=1}^{k}(a_i) = s`?
+
+        This function solves produces a list of all sequences which satisfy this constraint:
```
Comment (AmintorDusko) [inline pennylane/resource/error/trotter_error.py:105]:
```suggestion
        This function produces a list of all sequences which satisfy this constraint:
```

File: pennylane/resource/error/trotter_error.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+r""" Private functions implementing the error analysis for Trotter-Suzuki product formulas."""
+
+import math
+from functools import reduce, lru_cache
+
+import pennylane as qml
+
+
+# General Helper functions
+def _compute_repetitions(order, n):
```
Comment (AmintorDusko) [inline pennylane/resource/error/trotter_error.py:23]:
```suggestion
def _compute_repetitions(order: int, n: int):
```

File: pennylane/resource/error/trotter_error.py [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+    for op in h_ops:
+        h_one_norm += _spectral_norm(op, fast=fast)
+
+    c = (h_one_norm * t) ** (p + 1) / (math.factorial(p + 1) * n**p)
+    return c * (upsilon ** (p + 1) + 1)
+
+
+# Compute alpha_comm
+@lru_cache
+def _generate_combinations(num_variables, required_sum):
```
Comment (AmintorDusko) [inline pennylane/resource/error/trotter_error.py:96]:
```suggestion
def _generate_combinations(num_variables: int, required_sum: int):
```

File: pennylane/resource/error/trotter_error.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+    master_lst = []
+    for i in range(required_sum + 1):
+        for sub_perms in _generate_combinations(num_variables - 1, required_sum - i):
+            master_lst.append((i,) + sub_perms)
+
+    return tuple(master_lst)
+
+
+@lru_cache
+def _recursive_nested_commutator(A, B, alpha):
```
Comment (AmintorDusko) [inline pennylane/resource/error/trotter_error.py:138]:
```suggestion
def _recursive_nested_commutator(A, B, alpha: int):
```

File: pennylane/resource/error/trotter_error.py [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+        return B
+
+    if alpha == 1:
+        return qml.comm(A, B)
+
+    return qml.comm(A, _recursive_nested_commutator(A, B, alpha - 1))
+
+
+# Compute commutator error:
+def _commutator_error(h_ops, t, p, n, fast):
```
Comment (AmintorDusko) [inline pennylane/resource/error/trotter_error.py:167]:
```suggestion
def _commutator_error(h_ops, t: float, p: int, n: int, fast: bool) -> float:
```

File: pennylane/resource/error/trotter_error.py [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+    r"""Compute an upper-bound on the spectral norm error for approximating
+    the time evolution of a hamiltonian using a Suzuki-Trotter product formula.
```
Comment (AmintorDusko) [inline pennylane/resource/error/trotter_error.py:169]:
```suggestion
    the time evolution of a Hamiltonian using a Suzuki-Trotter product formula.
```

File: pennylane/resource/error/trotter_error.py [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+
+                nested_comm = _recursive_nested_commutator(H_i, nested_comm, alpha_i)
+
+            h_comm_norm += _spectral_norm(nested_comm, fast=fast) / c
+
+    return pre_factor * h_comm_norm
+
+
+# Flatten the product formula
+def _recursive_flatten(order, num_ops, scalar_t):
```
Comment (AmintorDusko) [inline pennylane/resource/error/trotter_error.py:211]:
```suggestion
def _recursive_flatten(order: int, num_ops: int, scalar_t: float):
```

File: pennylane/resource/error/trotter_error.py [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+    r"""Constructs a flattened list representation of the Suzuki-Trotter product formula.
+
+    Suppose we have a hamiltonian :math:`H = 1.2*X + 0.5*Y`, an associated 2nd order product formula is:
```
Comment (AmintorDusko) [inline pennylane/resource/error/trotter_error.py:214]:
```suggestion
    Suppose we have a Hamiltonian :math:`H = 1.2*X + 0.5*Y`, an associated 2nd order product formula is:
```

File: pennylane/resource/error/trotter_error.py [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+        (2 * coeff_lst_1) + coeff_lst_2 + (2 * coeff_lst_1),
+    )
+
+
+def _simplify(ops_index, coeffs):
+    r"""A helper function which collects like terms and combines coefficients to return a simplified product.
+
+    **Example:**
+
+        Suppose we have a hamiltonian :math:`H = 1.2*X + 0.5*Y`, an associated 2nd order product formula is:
```
Comment (AmintorDusko) [inline pennylane/resource/error/trotter_error.py:261]:
```suggestion
        Suppose we have a Hamiltonian :math:`H = 1.2*X + 0.5*Y`, an associated 2nd order product formula is:
```

File: pennylane/resource/error/trotter_error.py [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+
+        if (i + 1 < iter_limit) and (ops_index[i] == ops_index[i + 1]):
+            final_coeffs[-1] += coeffs[i + 1]
+            shift = 2
+
+        i += shift
+    return (final_ops, final_coeffs)
+
+
+def _flatten_trotter(num_ops, order, n):
```
Comment (AmintorDusko) [inline pennylane/resource/error/trotter_error.py:307]:
```suggestion
def _flatten_trotter(num_ops: int, order: int, n: int):
```

File: pennylane/resource/error/trotter_error.py [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+    coefficients. :code:`ops = [X(0), Y(0), Y(0), X(0)], coeffs = [1.2/2, 0.5/2, 0.5/2, 1.2/2]`. We can
+    further compress the memory needed to store the product formula by simply storing the indicies of the operators as
+    they appear in the hamiltonian. Since :math:`H = 1.2*X + 0.5*Y`, we have :code:`ops_index = [0, 1, 1, 0]`.
+
+    The Suzuki-Trotter product formula is defined recursively (see :class:`~.TrotterProduct`). This function recursively
+    constructs the flattened list representation for the product formula.
+
+    Args:
+        order (int): The order of the product formula.
+        num_ops (int): The number of terms in the hamiltonian.
```
Comment (AmintorDusko) [inline pennylane/resource/error/trotter_error.py:230]:
```suggestion
        num_ops (int): The number of terms in the Hamiltonian.
```

File: pennylane/templates/subroutines/trotter.py [hunk 14]
```diff
@@ -207,6 +207,64 @@ def __init__(  # pylint: disable=too-many-arguments
         }
         super().__init__(time, wires=hamiltonian.wires, id=id)
 
+    def error(self, method="commutator", fast=True):  # pylint: disable=arguments-differ
```
Comment (AmintorDusko) [inline pennylane/templates/subroutines/trotter.py:210]:
```suggestion
    def error(self, method: str = "commutator", fast: = True):  # pylint: disable=arguments-differ
```

**Review by AmintorDusko — APPROVED** [review]
Nice work! I left some nitpick suggestions, but feel free to ignore if sensible.

**Review by DSGuala — COMMENTED** [review]
Looks good. Left a few small comments. I'm confused about the scaling of the error by changes in order and number of steps.

File: pennylane/templates/subroutines/trotter.py [hunk 15]
```diff
@@ -207,6 +207,64 @@ def __init__(  # pylint: disable=too-many-arguments
...
+        >>> op = qml.TrotterProduct(hamiltonian, time=0.01, order=2)
+        >>> op.error(method="one-norm")
+        SpectralNormError(8.039062500000003e-06)
+
+        The "commutator" error bound can be computed by passing the kwarg :code:`method="commutator"`, and
+        is defined according to Appendix C (equation 189) `Childs et al. (2021) <https://arxiv.org/abs/1912.08854>`_.
+
+        >>> hamiltonian = qml.dot([1.0, 0.5, -0.25], [qml.X(0), qml.Y(0), qml.Z(0)])
+        >>> op = qml.TrotterProduct(hamiltonian, time=0.01, order=2)
+        >>> op.error(method="commutator")
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/templates/subroutines/trotter.py:239]: Fixed! 
image

File: pennylane/templates/subroutines/trotter.py [hunk 16]
```diff
@@ -207,6 +208,64 @@ def __init__(  # pylint: disable=too-many-arguments
         }
         super().__init__(time, wires=hamiltonian.wires, id=id)
 
+    def error(self, method="commutator", fast=True):  # pylint: disable=arguments-differ
+        # pylint: disable=protected-access
+        r"""Compute an upper-bound on the spectral norm error for approximating the
```
Comment (Jaybsoni) [inline pennylane/templates/subroutines/trotter.py:213]:
```suggestion
        r"""Compute an *upper-bound* on the spectral norm error for approximating the
```

File: pennylane/templates/subroutines/trotter.py [hunk 17]
```diff
@@ -207,6 +207,64 @@ def __init__(  # pylint: disable=too-many-arguments
...
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/templates/subroutines/trotter.py:213]: Added

File: tests/templates/test_subroutines/test_trotter.py [hunk 18]
```diff
@@ -23,8 +23,10 @@
 import pennylane as qml
 from pennylane import numpy as qnp
 from pennylane.math import allclose, get_interface
+from pennylane.resource.error import SpectralNormError
 from pennylane.templates.subroutines.trotter import _recursive_expression, _scalar
 
+
```
Comment (Jaybsoni) [inline tests/templates/test_subroutines/test_trotter.py:29]:
```suggestion
```

File: pennylane/resource/error/trotter_error.py [hunk 19]
```diff
@@ -0,0 +1,322 @@
...
+
+        if (i + 1 < iter_limit) and (ops_index[i] == ops_index[i + 1]):
+            final_coeffs[-1] += coeffs[i + 1]
+            shift = 2
+
+        i += shift
+    return (final_ops, final_coeffs)
+
+
+def _flatten_trotter(num_ops, order):
```
Comment (Jaybsoni) [inline pennylane/resource/error/trotter_error.py:307]:
```suggestion
def _flatten_trotter(num_ops: int, order: int):
```

File: pennylane/resource/error/trotter_error.py [hunk 20]
```diff
@@ -0,0 +1,322 @@
...
+
+import math
+from functools import reduce, lru_cache
+
+import pennylane as qml
+
+
+# General Helper functions
+def _compute_repetitions(order):
+    r"""Count the number of times the hamiltonian is repeated ("stages")
```
Comment (Jaybsoni) [inline pennylane/resource/error/trotter_error.py:24]:
```suggestion
    r"""Count the number of times the Hamiltonian is repeated ("stages")
```

File: pennylane/resource/error/trotter_error.py [hunk 21]
```diff
@@ -0,0 +1,322 @@
...
+r""" Private functions implementing the error analysis for Trotter-Suzuki product formulas."""
+
+import math
+from functools import reduce, lru_cache
+
+import pennylane as qml
+
+
+# General Helper functions
+def _compute_repetitions(order):
```
Comment (Jaybsoni) [inline pennylane/resource/error/trotter_error.py:23]:
```suggestion
def _compute_repetitions(order: int):
```

File: pennylane/templates/subroutines/trotter.py [hunk 22]
```diff
@@ -61,7 +62,7 @@ def _recursive_expression(x, order, ops):
     return (2 * ops_lst_1) + ops_lst_2 + (2 * ops_lst_1)
 
 
-class TrotterProduct(Operation):
+class TrotterProduct(ErrorOperation):
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/templates/subroutines/trotter.py:65]: Added !

File: tests/templates/test_subroutines/test_trotter.py (hunk 1 above)
Comment (DSGuala) [reply tests/templates/test_subroutines/test_trotter.py:520]: If it doesn't break TrotterProduct and only affects the error, I'm okay with leaving this as tech debt for Q2.

**Review by DSGuala — APPROVED** [review]
Looks good to me :+1:

**Comment by Jaybsoni** [comment]
It doesn't break TortterProduct, Its just not compatible with the error computation. I added a warning to make this clear to users.

File: tests/templates/test_subroutines/test_trotter.py (hunk 1 above)
Comment (Jaybsoni) [reply tests/templates/test_subroutines/test_trotter.py:520]: It doesn't break TortterProduct, Its just not compatible with the error computation. I added a warning to make this clear to users.

File: pennylane/templates/subroutines/trotter.py [hunk 23]
```diff
@@ -132,6 +133,17 @@ def my_circ():
     .. details::
         :title: Usage Details
 
+        An *upper-bound* for the error in approximating time-evolution using this operator can be
+        computed by calling the error function. It is computed using two different methods; the
```
Comment (soranjh) [inline pennylane/templates/subroutines/trotter.py:137]: Please change to :func:`~.TrotterProduct.error()` or something similar.

File: pennylane/templates/subroutines/trotter.py [hunk 24]
```diff
@@ -132,6 +133,17 @@ def my_circ():
...
+        "one-norm" scaling method and the "commutator" scaling method. (see `Childs et al. (2021) <https://arxiv.org/abs/1912.08854>`_.)
```
Comment (soranjh) [inline pennylane/templates/subroutines/trotter.py:138]:
```suggestion
        ``"one-norm"`` scaling method and the ``"commutator"`` scaling method. (see `Childs et al. (2021) <https://arxiv.org/abs/1912.08854>`_)
```

File: pennylane/resource/error/trotter_error.py [hunk 25]
```diff
@@ -0,0 +1,322 @@
...
+    Returns:
+        float: An upper-bound for the spectral norm error of the product formula.
+    """
+    upsilon = _compute_repetitions(p)
+    h_one_norm = 0
+
+    for op in h_ops:
+        h_one_norm += _spectral_norm(op, fast=fast)
+
+    c = (h_one_norm * t) ** (p + 1) / (math.factorial(p + 1) * n**p)
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/resource/error/trotter_error.py:90]: Yes, this comes from this equation in the paper: 

image

File: pennylane/resource/error/trotter_error.py [hunk 26]
```diff
@@ -0,0 +1,322 @@
...
+        t (float): The time interval for evolution.
+        p (int): The order of the product formula.
+        n (int): The number of Trotter steps (repetitions).
+        fast (bool): If True, a Frobenius bound is used to approximate the spectral norms of each term.
+
+    Returns:
+        float: An upper-bound for the spectral norm error of the product formula.
+    """
+    upsilon = _compute_repetitions(p)
+    pre_factor = (2 * upsilon * t ** (p + 1)) / ((p + 1) * n**p)
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/resource/error/trotter_error.py:186]: yes, we pick up a factor of p! in the numerator from the nested commutators. The n^p is from similar analysis as above. 
image

File: pennylane/templates/subroutines/trotter.py (hunk 23 above)
Comment (Jaybsoni) [reply pennylane/templates/subroutines/trotter.py]: Added

**Pull request merged**